app/src/main/java/com/github/siela1915/bootcamp/UploadingRecipeFragment.java

```diff
@@ -155,6 +180,63 @@ public void onClick(View v) {
         return view;
     }
 
+
+    public void setupIngredientAutocomplete(AutoCompleteTextView ingredientAutoComplete, Map<String, Integer> idMap, IngredientAutocomplete apiService){
+        ingredientAutoComplete.setThreshold(1);
+        //clear the field if the user wants to re-enter the ingredient name
+        ingredientAutoComplete.setOnFocusChangeListener(new View.OnFocusChangeListener() {
+            @Override
+            public void onFocusChange(View v, boolean hasFocus) {
+                if(hasFocus){
+                    ingredientAutoComplete.setText("");
+                }
+            }
+        });
+        ingredientAutoComplete.addTextChangedListener(new TextWatcher() {
+            String prevString;
+            boolean isTyping = false;
+            @Override
+            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
+            }
+            @Override
+            public void onTextChanged(CharSequence s, int start, int before, int count) {
+            }
+            private Timer timer = new Timer();
+            private final long DELAY = 1000; // milliseconds
+            @Override
+            public void afterTextChanged(final Editable s) {
+                //couldn't figure out a way to test an api inside UI classes
+                if(BuildConfig.DEBUG){
+                    ArrayAdapter<String> ingredientAdapter = new ArrayAdapter<String>(ingredientAutoComplete.getContext(), android.R.layout.select_dialog_item, Arrays.asList("apple"));
+                    ingredientAutoComplete.setAdapter(ingredientAdapter);
+                    ingredientAutoComplete.showDropDown();
+                    return;
+                }
+                //doesn't consider defocusing and refocusing the text field as typing
+                if(!s.toString().equals(prevString)){
+                    if (!isTyping) {
+                        // Send notification for start typing event
+                        isTyping = true;
+                    }
+                    timer.cancel();
+                    timer = new Timer();
+                    timer.schedule(
+                            new TimerTask() {
+                                @Override
+                                public void run() {
+                                    isTyping = false;
+                                    prevString = s.toString();
+                                    //send notification for stopped typing event
+                                    apiService.completeSearchNames(s.toString(), ingredientAutoComplete, idMap);
+                                }
+                            },
+                            DELAY
+                    );
```

Comment:
The code looks good enough for now! In the future, the timer used to delay API calls could be replaced with a handler to simplify the code. Also, it would be better to cancel the previous API call if the user continues to type, as the current code could result in multiple calls to the same API if the user types fast. :)